[Back to Main Menu](https://github.com/sherman6/tutorials/blob/master/README.md#menu)  

<a class="anchor" id="topmenu"></a>

# Python Tutorial 03:

# Writing Useful Functions in Python


___

_Sherman6,  2020 February_

___


### Key Concepts Covered:  


* __[The Basics](#first-bullet)__
    * A quick rundown of some function basics  
    

* __[Global Variables](#second-bullet)__
    * The difference between global and local variables, often the answer to: 'Why doesn't my function work?' 
    
    
* __[How to Define Global Variables](#third-bullet)__
    * How to define global variables, with plenty of examples. 
    

* __[Using Functions in Coordination ("Useful" Functions)](#seventh-bullet)__
    * Functions are most-useful when used in coordination with each other.  Here's how to call a function from another function  


* __[Appendix](#appendix)__
    * References, More Resources, & Machine Information 
    
___ 

### Introduction:


Python's popularity continues to grow, with it becoming the #1 language preferred by those holding data science positions in 2019, according to Burtch Works [[1]](#reference).  This is not without good reason: Python is a great language for connecting disparate analytics tools to form a unified production pipeline, with benefits including ease of use, interpretability, an active user community, and the wide array of libraries available for various purposes, the quantity of which continues to grow [[2]](#reference).  And, of course, _it's free!_  


___

### Purpose:

#### In each of these tutorials, I will demonstrate a few tips and tricks which will hopefully help others troubleshoot code in Python.  

___


One of Python's benefits is that there are often many different ways to achieve something.  However, there are often _even more ways_ one can get stuck trying to figure it out.  

In my experience, seeing examples of __"the wrong way"__ to code is just as educational as seeing __"the right way"__ to code, so in several places I show examples of things that DON'T work, the error messages resulting from them, and why those messages appear. 

Furthermore, I've spent my fair share of hours [Googling](https://www.google.com), [Stack Overflow-ing](https://stackoverflow.com/), and [GitHub-ing](https://github.com/) to find solutions, as well as a decent amount of time on some very informative and robust free tutorial websites, such as [W3Schools](https://www.w3schools.com/), [TutorialsPoint](https://www.tutorialspoint.com/python/index.htm), and [R-Bloggers](https://www.r-bloggers.com/) (the first that come to mind).  
- While there are some great documentation and tutorials out there, I find myself __jumping from one to the next__, until I have __dozens of browser tabs open__, as each focuses _very in-depth_ on a specific command, library, function, or variable type.  


- That is why I'm designing these tutorials to be __broad, rather than deep__.  These tutorials are __not designed to be exhaustive__, but rather to help troubleshoot when you get stuck, and share a few helpful tips along the way. 



A few final comments:

- This workbook assumes at least basic familiarity with Python 3 and Jupyter Notebooks.

- More advanced topics, such as in-depth examples of DataFrames, and specific nuances for neural network training, will be covered in separate workbooks.  

___

_Version: 02_   

In [ ]:
#Importing packages:
#import numpy as np
#import pandas as pd
#mport random
#import re

<a class="anchor" id="first-bullet"></a>
___
___
___

## The Basics
[Back to Top Menu](#topmenu)  



User-defined functions (also known as udf's) have many benefits, such as:  

- Reducing the verbosity of your code (making it shorter easier to read) 
- Enhanced scaleability (doing the same thing many, many times)
- Versatility (ability to handle different scenarios without writing specific commands for each) 
- Accomplishing more, with less (this, basically)


Let's go over a few basics, first.  __This is by no means comprehensive.__  

_For any python basic troubleshooting, such as working with lists or dataframes, please see Tutorial 1; these will not be described here_.  

___


Let's define a simple function.  

In [1]:
def myfunction(input):
    output = input + 2
    return output

In [2]:
myfunction(3)

5

Here's another:

In [3]:
def concatenate_two_things(input1, input2):
    output = input1+" "+"&"+" "+input2
    return output

In [4]:
concatenate_two_things("Harry", "Sally")

'Harry & Sally'

In [5]:
concatenate_two_things("Peanut Butter", "Jelly")

'Peanut Butter & Jelly'

In [6]:
concatenate_two_things("Bert", "Ernie")

'Bert & Ernie'

As you can see, there are 3 key parts:
- `def _____:` : Where you define the name of the function, the inputs it expects to receive, any default parameters, et cetera. 


- The inner section, where you define the function.  This can be as big or small as you wish. Remember: Indent levels matter!


- `return` : Where you specify what you want to be returned from that function.  You could call this the "output", but don't be misled - a function can function (pun) even if this is not specified.  More on that, below.  

For example, here's what would happen if you __didn't__ specify a `return` argument in the above function. 


In [7]:
x = 10

In [8]:
def myfunction(input):
    input + 2

In [9]:
print(myfunction(x))
print(x)

None
10


Nothing happened. 

In [10]:
z = myfunction(3)
print(z)

None


Nothing happened. 

As mentioned, a function __can still__ generate an output & perform an action on an object even if you don't specify it to 'return' anything, of course, but that leads us into the next topic. 


<a class="anchor" id="second-bullet"></a>
___
___
___

## Global Variables
[Back to Top Menu](#topmenu)  



That's right, variables can be __worldwide__ (like [Pitbull](https://open.spotify.com/playlist/5HOXE1APnmH5hFvUM35YH5)).  

This is distinguished from __"local"__ variables that __only exist within the function itself__.  Local variables are temporary - they only exist while the function is acting, and cannot be retrieved by other functions after it ceases. 

- That is a shame, because look how little time our earlier function takes to complete:

In [11]:
%time
myfunction(1)

CPU times: user 2 µs, sys: 1 µs, total: 3 µs
Wall time: 3.58 µs


- 3 µs is pretty fast.  My point is, 'output' is created, then 'poof', it's gone.  

___  

Local variables can only be accessed by the function within which they are defined.  

This means that you can't call a local variable of function A from within function A.  

In [12]:
def function(b):
    a = 4
    sumvariable = a+b
    return sumvariable

In [13]:
function(10)

14

In the above function, 
- We assign 'b' as 10. 
- 'a' is a local variable. 

In [14]:
sumvariable

NameError: name 'sumvariable' is not defined

Takeaway: __it doesn't exist__.

In [15]:
a

NameError: name 'a' is not defined

As you can see, this variable doesn't exist outside of it's (very) temporary existence within the function.  

___
Similarly, local variables cannot overwrite global ones with the same name. 

In [16]:
a=5
print(a)

5


In [17]:
def function(input):
    a = 100
    sumvariable = input+a
    return sumvariable

In [18]:
function(1)

101

In [19]:
print(a)

5


As we can see, when we print 'a', we get the object that we defined prior to writing the function (a=5), NOT the value 'a=100', which was defined within the function itself. This is because in the above example, the object 'a' within the function is NOT global, while the object 'a' outside of the function IS global. 

Another example:

In [20]:
sumvariable="parameter 1"
print(sumvariable)

parameter 1


In [21]:
def function(input):
    a = 100
    sumvariable = input+a
    return sumvariable

In [22]:
function(1)

101

In [23]:
print(sumvariable)

parameter 1


<a class="anchor" id="third-bullet"></a>
___
___
___

## How to Define Global Variables:  
[Back to Top Menu](#topmenu)  



In [24]:
def function(input):
    global sumvariable
    a = 100
    sumvariable = input+a
    return sumvariable

In [25]:
function(1)

101

In [26]:
print(sumvariable)

101


We just created a new global variable, `sumvariable`, by adding the word 'global' in front of it, prior to performing any action on it or assigning it.  

It not only exists outside the function, but it will overwrite any object of the same name that previously existed outside of the function (if it is assigned within the function). 

- This becomes very important when using nested functions, or functions within functions, or otherwise having a function call another function.  

#### Example WITH 'global':

In [27]:
sumvariable = 10

In [28]:
def function(input):
    global sumvariable
    a = 100
    sumvariable = input+a
    return sumvariable

In [29]:
function(sumvariable)

110

In [30]:
print(sumvariable)

110


As we can see, the global object 'sumvariable', which originally was '10', took on a new value (110) after being modified by the function (where it was named as a global variable). 

#### Example WITHOUT 'global':

In [31]:
sumvariable = 10

In [32]:
def function(input):
    a = 100
    sumvariable = input+a
    return sumvariable

In [33]:
function(sumvariable)

110

In [34]:
print(sumvariable)

10


In the above example, the value of the global object 'sumvariable' never changed from it's original value of '10', because the object 'sumvariable' which existed onlyy within the function was not specified as being 'global'.  

#### One last example:  

In [35]:
def function(b):
    global c
    global d
    global e
    c=123
    d="I exist!"
    e=[]
    sumvar = 4+b
    return sumvar

In [36]:
function(5)

9

In [37]:
c

123

In [38]:
d

'I exist!'

In [39]:
e

[]

As we can see, our global variable `e` is blank, but it DOES exist (having been created when we ran the line of code, `function(5)` ).

<a class="anchor" id="seventh-bullet"></a>
___
___
___

## Using Functions in Coordination ("Useful" Functions)

[Back to Top Menu](#topmenu)  


Let's review some of the key benefits of using functions:  


- Reducing the verbosity of your code  
- Enhanced scaleability  
- Versatility 
- Accomplishing more, with less 

In my experience, functions are most-useful (that is, most-helpful at acheiving the aforemented goals) when they are used __in coordination with each other__.



___

In [40]:
del(sumvariable) #delete it to wipe it off the environment, first. 
#now, it no longer exists.

In [41]:
def function(input):
    global sumvariable
    sumvariable = input+100
    return sumvariable

In [42]:
function(1)

101

In [43]:
def func2(x):
    return x+sumvariable

In [44]:
func2(1000)

1101

As we can see, we can now call the output of one function from another function.  

___

As mentioned, we don't need to specify __return__, in this case. 

In [45]:
del(sumvariable) #delete it to wipe it off the environment, first. 
#now, it no longer exists.

In [46]:
def function(input):
    global sumvariable
    sumvariable = input+100

In [47]:
function(1)

In [48]:
sumvariable

101

In [49]:
def func2(x):
    return function(x)

In [50]:
func2(1000)

In [51]:
sumvariable

1100

<a class="anchor" id="appendix"></a>
___

# Appendix:

[Back to Top Menu](#topmenu)  


___

<a class="anchor" id="reference"></a>

### References:

[1] https://www.burtchworks.com/2019/08/21/2019-sas-r-or-python-survey-update-which-tool-do-data-scientists-analytics-pros-prefer/

[2]  https://www.burtchworks.com/2019/12/10/metis-data-scientists-on-pythons-advantages-growing-popularity/


### More Resources:

- https://www.tutorialspoint.com/python/python_functions.htm
    - A pretty thorough tutorial on defining functions in Python. 
- http://anh.cs.luc.edu/python/hands-on/3.1/handsonHtml/functions.html
    - A helpful and also very thorough hands-on tutorial for writing functions on your own, from the Computer Science Department at Loyola University Chicago. 
- http://anh.cs.luc.edu/python/hands-on/3.1/handsonHtml/index.html
    - The full, very thorough and deep hands-on Python 3 tutorial from the Computer Science Department at Loyola University Chicago. 


### Machine Information:  

This workbook was run on... 

In [52]:
#Python Version:
import sys
print(sys.version)

3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0]


In [53]:
#Timestamp:
import datetime
datetime.datetime.now().strftime("%a, %d %B %Y %H:%M:%S")

'Fri, 28 February 2020 19:25:19'

In [54]:
#Operating System:
import os
print(os.name)
print(sys.platform)

posix
linux



___

_Sherman6,  2020 February_

___
